modules/tensor_mechanics/src/constraints/NodalFrictionalConstraint.C

```diff
@@ -0,0 +1,342 @@
+  _connected_nodes.clear();
+  _master_conn.clear();
+  for (unsigned int j = 0; j < slave_nodelist.size(); ++j)
+  {
+    if (_mesh.nodeRef(slave_nodelist[j]).processor_id() == _subproblem.processor_id())
+    {
+      Node & slave_node = _mesh.nodeRef(slave_nodelist[j]);
+      for (unsigned int i = 0; i < _master_node_vector.size(); ++i)
+      {
+        Node & master_node = _mesh.nodeRef(_master_node_vector[i]);
+        Real d = (slave_node - master_node).norm();
+        if (MooseUtils::absoluteFuzzyEqual(d, 0.0))
+        {
+          _master_conn.push_back(i);
+          _connected_nodes.push_back(slave_nodelist[j]);
+          break;
+        }
+      }
+    }
+    /*if (_master_conn[j] == std::numeric_limits<unsigned int>::max())
+    {
+        printf("in here \n");
+        mooseError("No master node located at the same elevation as the slave node.");
+    }*/
+  }
+  printf("total slave nodes, master nodes: %lu, %lu \n",
+         _master_conn.size(),
+         _master_node_vector.size());
+}
+
+void
+NodalFrictionalConstraint::computeResidual(NumericVector<Number> &
+                                           /*residual*/)
+{
+  std::vector<dof_id_type> masterdof = _var.dofIndices();
+  std::vector<dof_id_type> slavedof = _var.dofIndicesNeighbor();
+  DenseVector<Number> re(masterdof.size());
+  DenseVector<Number> neighbor_re(slavedof.size());
+
+  re.zero();
+  neighbor_re.zero();
+
+  for (_i = 0; _i < slavedof.size(); ++_i)
+  {
+    _j = _master_conn[_i];
+    switch (_formulation)
+    {
+      case Moose::Penalty:
+        re(_j) += computeQpResidual(Moose::Master);
+        neighbor_re(_i) += computeQpResidual(Moose::Slave);
+        break;
+      case Moose::Kinematic:
+        mooseError("This constraint currently works only with penalty formulation.");
+    }
+  }
+  _assembly.cacheResidualNodes(re, masterdof);
+  _assembly.cacheResidualNodes(neighbor_re, slavedof);
+}
+
+Real
+NodalFrictionalConstraint::computeQpResidual(Moose::ConstraintType type)
+{
+  // check whether the tangential spring is already in the yielded state
+  Real old_force = (_u_slave_old[_i] - _u_master_old[_j]) * _tangential_penalty;
+  if (MooseUtils::absoluteFuzzyGreaterThan(std::abs(old_force),
+                                           _friction_coefficient * _normal_force))
+    old_force = _friction_coefficient * _normal_force * old_force / std::abs(old_force);
+
+  Real current_force = ((_u_slave[_i] - _u_slave_old[_i]) - (_u_master[_j] - _u_master_old[_j])) *
+                           _tangential_penalty +
+                       old_force;
+  printf("current, max: %e, %e\n", std::abs(current_force), _friction_coefficient * _normal_force);
+  if (MooseUtils::absoluteFuzzyGreaterThan(std::abs(current_force),
+                                           _friction_coefficient * _normal_force))
+  {
+    current_force = _friction_coefficient * _normal_force * current_force / std::abs(current_force);
+    printf("yielded: %u \n", _i);
```

Comment:
This one too.  Also, get rid of the curly braces around this one-line block.

Reply:
Done.